// github-org-stats/lib/GithubOrgStats.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _github = require('github');

var _github2 = _interopRequireDefault(_github);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var GithubOrgStats = function GithubOrgStats(token, org) {
  var _this = this;

  _classCallCheck(this, GithubOrgStats);

  this.getOrgRepos = function () {
    return _this.github.repos.getForOrg({ org: _this.org, per_page: 100 });
  };

  this.get = function () {
    return _this.getOrgRepos().then(function (_ref) {
      var data = _ref.data;

      var repos = data.map(function (repo) {
        return { name: repo.name, stars: parseInt(repo.stargazers_count, 10) };
      }).sort(function (a, b) {
        return b.stars - a.stars;
      });

      var totalStars = repos.reduce(function (total, current) {
        return total + current.stars;
      }, 0);

      return {
        totalStars: totalStars,
        repos: repos
      };
    });
  };

  this.token = token;
  this.org = org;

  this.github = (0, _github2.default)();

  this.github.authenticate({
    type: 'token',
    token: token
  });
};

exports.default = GithubOrgStats;
